network/phonebook.go

```diff
@@ -82,7 +105,48 @@ func (e *phonebookEntries) ReplacePeerList(they []string) {
 }
 
 func (e *phonebookEntries) updateRetryAfter(addr string, retryAfter time.Time) {
-	(*e)[addr] = phonebookData{retryAfter: retryAfter}
+	_, found := (*e)[addr]
+	if !found {
+		(*e)[addr] = phonebookData{retryAfter: retryAfter, recentConnectionTimes: make([]time.Time, 0)}
+	} else {
+		phbData := (*e)[addr]
+		phbData.retryAfter = retryAfter
+		(*e)[addr] = phbData
+	}
+}
+
+// waitAndAddConnectionTime will wait to prevent exceeding connectionsRateLimitingCount.
+// Then it will register the next connection time.
+func (e *phonebookEntries) waitAndAddConnectionTime(addr string,
+	connectionsRateLimitingCount uint, connectionsRateLimitingWindowSeconds uint) {
+	connectionRateLimitWindowDuration := time.Duration(connectionsRateLimitingWindowSeconds)*time.Second
+	_, found := (*e)[addr]
+	if !found {
+		(*e)[addr] = phonebookData{retryAfter: time.Now(), recentConnectionTimes: make([]time.Time, 0)}
+	}
+
+	var timeSince time.Duration
+	// Remove from recentConnectionTimes the times later than ConnectionsRateLimitingWindowSeconds
+	for len((*e)[addr].recentConnectionTimes) > 0 {
+		timeSince = time.Since(((*e)[addr].recentConnectionTimes)[0])
```

Comment:
Using time.Since is abit of an issue since the time is moving forward while the function is executing. You want to get a reference time ( i.e. `curTime := time.Now()` ) and use that all across the function.
Also, in this section you want to find the "tail" and cut it in a single slice rather than one item at a time.